README.md

# <img src="man/figures/bento-logotype-singleline-black.png" width="400px" style="background-color:white;border=transparent" /> <img src="man/figures/logo.png" align="right" width="140px" style="padding-left:20px; background-color:white" />

<!-- badges: start -->
[![R build status](https://github.com/PhanstielLab/BentoBox/workflows/R-CMD-check-bioc/badge.svg)](https://github.com/PhanstielLab/BentoBox/actions)
<!-- badges: end -->

## Overview

`BentoBox` is a genomic data visualization package for R. Using `grid`
graphics, `BentoBox` empowers users to programmatically and flexibly
generate multi-panel figures. `BentoBox` accomplishes these goals by
utilizing 1) a coordinate-based plotting system, and 2) edge-to-edge
containerized data visualization. The coordinate-based plotting system
grants users precise control over the size, position, and arrangement of
plots. Its edge-to-edge plotting functions preserve the mapping between
user-specified containers and the represented data. This allows users to
stack plots with confidence that vertically aligned data will correspond
to the same regions. For more information about BentoBox’s philosophy
and design, check out the `Our Philosophy` page.

Specialized for genomic data, `BentoBox` also contains functions to read
and plot multi-omic data quickly and easily. These functions are
integrated with Bioconductor packages to flexibly accommodate a large variety 
of genomic assemblies. `BentoBox` can address an
endless number of use cases, including: dynamic exploration of genomic
data, arrangment into multi-omic layouts, and survey plotting for
quickly viewing data across the genome. Check out our `vignettes` for
detailed examples and suggested use cases!

## Installation

BentoBox can be installed from GitHub as follows:

``` r
if (!requireNamespace("BiocManager", quietly = TRUE))
    install.packages("BiocManager")

if (!requireNamespace("remotes", quietly = TRUE))
    BiocManager::install("remotes")

remotes::install_github("PhanstielLab/BentoBox")
packageVersion("BentoBox")
```

Cytoband annotation datasets and example datasets and files are included
with the package BentoBoxData:

``` r
remotes::install_github("PhanstielLab/BentoBoxData")
```

## Usage

<img src="man/figures/homePage-1.png" width="672" style="display: block; margin: auto;" />

``` r
## Load libraries and datasets
library("BentoBox")
library("org.Hs.eg.db")
library("TxDb.Hsapiens.UCSC.hg19.knownGene")
library("BentoBoxData")
data("cytoBand.Hsapiens.UCSC.hg19")
data("bb_gmHicData")
data("bb_imrHicData")
data("bb_gmCTCFData")
data("bb_imrCTCFData")
data("bb_imrH3K27acData")
data("bb_gmH3K27acData")

## Create a BentoBox page
bb_pageCreate(width = 7, height = 4.25, default.units = "inches")

##########################
######## Panel A #########
##########################
## Text section label
bb_plotText(label = "A", fontsize = 12,
            x = 0.25, y = 0.25, just = "left", default.units = "inches")
## Set genomic and dimension parameters in a `bb_params` object
params_a <- bb_params(chrom = "chr21", chromstart = 28000000, chromend = 30300000, 
                      assembly = "hg19",
                      x = 0.25, width = 2.75, just = c("left", "top"), default.units = "inches")
## Double-sided Hi-C Plot
hicPlot_top <- bb_plotHicSquare(data = bb_gmHicData, params = params_a,
                                zrange = c(0, 200), resolution = 10000,
                                half = "top",
                                y = 0.5, height = 2.75)
hicPlot_bottom <- bb_plotHicSquare(data = bb_imrHicData, params = params_a,
                                   zrange = c(0, 70), resolution = 10000,
                                   half = "bottom",
                                   y = 0.5, height = 2.75)
## Annotate Hi-C heatmap legends
bb_annoHeatmapLegend(plot = hicPlot_bottom, fontsize = 7, 
                     x = 3.05, y = 0.5,
                     width = 0.07, height = 0.5, just = c("left", "top"),
                     default.units = "inches")
bb_annoHeatmapLegend(plot = hicPlot_top, fontsize = 7, 
                     x = .125, y = 0.5,
                     width = 0.07, height = 0.5, just = c("left", "top"),
                     default.units = "inches")
## Plot gene track
genes_a <- bb_plotGenes(params = params_a, stroke = 1, fontsize = 6,
                        y = 3.35, height = 0.4)
## Annotate genome label
bb_annoGenomeLabel(plot = genes_a, params = params_a, 
                   scale = "Mb", fontsize = 7,
                   y = 3.85)
##########################
######## Panel B #########
##########################
## Text section label
bb_plotText(label = "B", fontsize = 12,
            x = 3.5, y = 0.25, just = "left", default.units = "inches")
## Plot ideogram
bb_plotIdeogram(chrom = "chr21", assembly = "hg19",
                x = 3.5, y = 0.5,
                width = 3.25, height = 0.15, just = c("left", "top"), default.units = "inches")
## Add text to ideogram
bb_plotText(label = "Chromosome 21", fontsize = 8, fontcolor = "darkgrey",
            x = 6.75, y = 0.4, just = "right", default.units = "inches")
##########################
######## Panel C #########
##########################
## Text section label
bb_plotText(label = "C", fontsize = 12,
            x = 3.5, y = 1, just = c("left", "top"), default.units = "inches")
## Set genomic and dimension parameters in a `bb_params` object
params_c <- bb_params(chrom = "chr21", chromstart = 28150000, chromend = 29150000, 
                      assembly = "hg19",
                      x = 3.5, width = 1.5, default.units = "inches")
## Set signal track data ranges
ctcf_range <- bb_params(range = c(0, 77),
                        assembly = "hg19")
hk_range <- bb_params(range = c(0, 32.6),
                      assembly = "hg19")
## Plot Hi-C triangle
hic_gm <- bb_plotHicTriangle(data = bb_gmHicData, params = params_c,
                             zrange = c(0, 200), resolution = 10000,
                             y = 1.75, height = 0.75, just = c("left", "bottom"))
## Annotate Hi-C heatmap legend
bb_annoHeatmapLegend(plot = hic_gm, fontsize = 7, 
                     x = 5, y = 1, width = 0.07, height = 0.5, 
                     just = c("right", "top"), default.units = "inches")
## Plot CTCF signal
ctcf_gm <- bb_plotSignal(data = bb_gmCTCFData, params = c(params_c, ctcf_range),
                         fill = "#253494", linecolor = "#253494",
                         y = 1.95, height = 0.6)
## CTCF label
bb_plotText(label = "CTCF", fontcolor = "#253494", fontsize = 8,
            x = 3.5, y = 1.95, just = c("left","top"), default.units = "inches")
## Plot H3K27ac signal
hk_gm <- bb_plotSignal(data = bb_gmH3K27acData, params = c(params_c, hk_range),
                       fill = "#37a7db", linecolor = "#37a7db",
                       y = 3.25, height = 0.6, just = c("left", "bottom"))
## H3K27ac label
bb_plotText(label = "H3K27ac", fontcolor = "#37a7db", fontsize = 8,
            x = 3.5, y = 2.65, just = c("left","top"), default.units = "inches")
## Plot genes
genes_gm <- bb_plotGenes(params = params_c, stroke = 1, fontsize = 6,
                         strandLabels = FALSE,
                         y = 3.35, height = 0.4)
## Annotate genome label
bb_annoGenomeLabel(plot = genes_gm, params = params_c, 
                   scale = "Kb", fontsize = 7,
                   y = 3.85)
##########################
######## Panel D #########
##########################
## Text section label
bb_plotText(label = "D", fontsize = 12,
            x = 5.25, y = 1, just = c("left", "top"), default.units = "inches")
## Set genomic and dimension parameters in a `bb_params` object
params_d <- bb_params(chrom = "chr21", chromstart = 28150000, chromend = 29150000, 
                      assembly = "hg19",
                      x = 6.75, width = 1.5, default.units = "inches")
## Plot Hi-C triangle
hic_imr <- bb_plotHicTriangle(data = bb_imrHicData, params = params_d,
                              zrange = c(0, 70), resolution = 10000,
                              y = 1.75, height = 0.75, just = c("right", "bottom"))
## Annotate Hi-C heatmap legend
bb_annoHeatmapLegend(plot = hic_imr, fontsize = 7,
                     x = 6.75, y = 1, width = 0.07, height = 0.5, just = c("right", "top"))
## Plot CTCF signal
ctcf_imr <- bb_plotSignal(data = bb_imrCTCFData, params = c(params_d, ctcf_range),
                          fill = "#253494", linecolor = "#253494",
                          y = 1.95, height = 0.6, just = c("right", "top"))
## Plot H3K27ac signal
hk_imr <- bb_plotSignal(data = bb_imrH3K27acData, params = c(params_d, hk_range),
                        fill = "#37a7db", linecolor = "#37a7db",
                        y = 3.25, height = 0.6, just = c("right", "bottom"))
## Plot gene track
genes_imr <- bb_plotGenes(params = params_d, stroke = 1,fontsize = 6,
                          strandLabels = FALSE,
                          y = 3.35, height = 0.4, just = c("right", "top"))
## Annotate genome label
bb_annoGenomeLabel(plot = genes_imr, params = params_d, 
                   scale = "Kb", fontsize = 7,
                   y = 3.85, just = c("right", "top"))
## Hide page guides
bb_pageGuideHide()
```

## A word of caution

`BentoBox` is incredibly flexible and functional. However, due to this
flexibility and like all programming packages, it may not always prevent
users from making unintentional mistakes. If plot sizes are entered
incorrectly or data is mishandled, it is possible to connect multi-omic
data incorrectly. Make sure you utilize package features that reduce
human error and increase re-usability of code to get the most milage out
of BentoBox.


NEWS
CHANGES IN VERSION 0.1.6
-------------------------

BUG FIXES 

    o R version requirement changed to (R >= 4.1.0) for proper plot placement.
    
NEW FEATURES

    o `colorby` object now has a `scalePerRegion` parameter to scale numerical
    `colorby` data to the range of data in a plotted genomic region.

CHANGES IN VERSION 0.1.5
-------------------------

SIGNIFICANT USER-VISIBLE CHANGES

    o `bb_plotManhattan` `fill` paramete now accepts a single value, 
    a vector of colors, or a `colorby` object.

CHANGES IN VERSION 0.1.4
-------------------------

SIGNIFICANT USER-VISIBLE CHANGES

    o `colorby` constructor now includes optional palette specification.
    o `bb_plotPairs`, `bb_plotPairsArches`, and `bb_plotRanges` `fill` parameter
    now accepts a single value, a vector of colors, or a `colorby` object.
    
BUG FIXES    

    o `GInteractions` assembly match checking moved before dataframe 
    conversion.

CHANGES IN VERSION 0.1.3
-------------------------

SIGNIFICANT USER-VISIBLE CHANGES

    o Data moved to `BentoBoxData` package.
    o Default genome assembly updated to "hg38".
    
BUG FIXES

    o Streamlined parameter parsing and data reading logic.

CHANGES IN VERSION 0.1.2
-------------------------

NEW FEATURES

    o Added unit tests with `testthat`.
    o `bb_annoDomains` function addition.
    o `bb_plotSignal` vertical orientation.
    
CHANGES IN VERSION 0.1.1
-------------------------

NEW FEATURES

    o Added a `NEWS` file to track changes to the package.
    
BUG FIXES

    o Updated viewport parsing for package `grid` version 4.1.0.

R/bb_plotRect.R
#' Plot a rectangle within a BentoBox layout
#' 
#' @usage bb_plotRect(
#'     x,
#'     y,
#'     width,
#'     height,
#'     just = "center",
#'     default.units = "inches",
#'     linecolor = "black",
#'     lwd = 1,
#'     lty = 1,
#'     fill = NA,
#'     alpha = 1,
#'     params = NULL,
#'     ...
#' )
#'
#' @param x A numeric vector or unit object specifying rectangle x-locations.
#' @param y A numeric vector, unit object, or a character vector of values
#' containing a "b" combined with a numeric value specifying
#' rectangle y-locations.
#' The character vector will place rectangle y-locations relative to
#' the bottom of the most recently plotted BentoBox plot according to
#' the units of the BentoBox page.
#' @param width A numeric vector or unit object specifying rectangle widths.
#' @param height A numeric vector or unit object specifying rectangle heights.
#' @param just Justification of rectangle relative to its (x, y) location.
#' If there are two values, the first value specifies horizontal justification
#' and the second value specifies vertical justification.
#' Possible string values are: \code{"left"}, \code{"right"},
#' \code{"centre"}, \code{"center"}, \code{"bottom"}, and \code{"top"}.
#' Default value is \code{just = "center"}.
#' @param default.units A string indicating the default units to use if
#' \code{x}, \code{y}, \code{width}, and \code{height} are only given as
#' numerics or numeric vectors.
#' Default value is \code{default.units = "inches"}.
#' @param linecolor A character value specifying rectangle line color.
#' Default value is \code{linecolor = "black"}.
#' @param lwd A numeric specifying rectangle line width.
#' Default value is \code{lwd = 1}.
#' @param lty A numeric specifying rectangle line type.
#' Default value is \code{lty = 1}.
#' @param fill A character value specifying rectangle fill color.
#' Default value is \code{fill = NA}.
#' @param alpha Numeric value specifying color transparency.
#' Default value is \code{alpha = 1}.
#' @param params An optional \link[BentoBox]{bb_params} object
#' containing relevant function parameters.
#' @param ... Additional grid graphical parameters. See \link[grid]{gpar}.
#'
#' @return Returns a \code{bb_rect} object containing
#' relevant placement and \link[grid]{grob} information.
#'
#' @examples
#' ## Create a BentoBox page
#' bb_pageCreate(width = 7.5, height = 6, default.units = "inches")
#'
#' ## Plot one rectangle with no fill
#' bb_plotRect(
#'     x = 0.5, y = 0.5, width = 3, height = 3,
#'     just = c("left", "top"), default.units = "inches",
#'     lwd = 2, fill = NA
#' )
#'
#'
#' ## Plot two rectangles with same width and height at different locations
#' bb_plotRect(
#'     x = 4, y = c(0.5, 2.25), width = 3, height = 1.25,
#'     just = c("left", "top"), default.units = "inches",
#'     fill = "#7ecdbb"
#' )
#'
#' ## Plot two rectangles with different widths, heights,
#' ## locations, and colors
#' bb_plotRect(
#'     x = 3.75, y = c(4, 5.25), width = c(6.5, 4.5),
#'     height = c(1, 0.25),
#'     just = "top", default.units = "inches",
#'     fill = c("#7ecdbb", "#37a7db"), linecolor = NA, alpha = 0.4
#' )
#'
#' ## Hide page guides
#' bb_pageGuideHide()
#' @seealso \link[grid]{grid.rect}
#'
#' @export
bb_plotRect <- function(x, y, width, height, just = "center",
                        default.units = "inches", linecolor = "black",
                        lwd = 1, lty = 1, fill = NA, alpha = 1,
                        params = NULL, ...) {

    # =========================================================================
    # PARSE PARAMETERS
    # =========================================================================

    bb_rectInternal <- parseParams(
        params = params,
        defaultArgs = formals(eval(match.call()[[1]])),
        declaredArgs = lapply(match.call()[-1], eval.parent, n = 2),
        class = "bb_rectInternal"
    )

    ## Set gp
    bb_rectInternal$gp <- gpar(
        col = bb_rectInternal$linecolor,
        fill = bb_rectInternal$fill,
        lwd = bb_rectInternal$lwd,
        lty = bb_rectInternal$lty,
        alpha = bb_rectInternal$alpha
    )
    bb_rectInternal$gp <- setGP(
        gpList = bb_rectInternal$gp,
        params = bb_rectInternal, ...
    )

    # =========================================================================
    # INITIALIZE OBJECT
    # =========================================================================

    bb_rect <- structure(list(
        x = bb_rectInternal$x, y = bb_rectInternal$y,
        width = bb_rectInternal$width,
        height = bb_rectInternal$height,
        just = bb_rectInternal$just,
        grobs = NULL, gp = bb_rectInternal$gp
    ),
    class = "bb_rect"
    )

    # =========================================================================
    # CATCH ERRORS
    # =========================================================================

    check_bbpage(error = "Cannot plot rectangle without a BentoBox page.")
    if (is.null(bb_rect$x)) {
        stop("argument \"x\" is missing, with no default.",
            call. = FALSE
        )
    }
    if (is.null(bb_rect$y)) {
        stop("argument \"y\" is missing, with no default.",
            call. = FALSE
        )
    }
    if (is.null(bb_rect$width)) {
        stop("argument \"width\" is missing, with no default.",
            call. = FALSE
        )
    }
    if (is.null(bb_rect$height)) {
        stop("argument \"height\" is missing, with no default.",
            call. = FALSE
        )
    }
    
    bb_checkColorby(fill = bb_rect$gp$fill,
                    colorby = FALSE)

    # =========================================================================
    # DEFINE PARAMETERS
    # =========================================================================

    ## Get page_height and its units from bbEnv through bb_makepage
    page_height <- get("page_height", envir = bbEnv)
    page_units <- get("page_units", envir = bbEnv)
    
    bb_rect$x <- misc_defaultUnits(
        value = bb_rect$x,
        name = "x",
        default.units = bb_rectInternal$default.units
    )
    bb_rect$y <- misc_defaultUnits(
        value = bb_rect$y,
        name = "y",
        default.units = bb_rectInternal$default.units
    )
    bb_rect$width <- misc_defaultUnits(
        value = bb_rect$width,
        name = "width",
        default.units = bb_rectInternal$default.units
    )
    bb_rect$height <- misc_defaultUnits(
        value = bb_rect$height,
        name = "height",
        default.units = bb_rectInternal$default.units
    )

    ## Convert coordinates to page_units
    new_x <- convertX(bb_rect$x, unitTo = page_units, valueOnly = TRUE)
    new_y <- convertY(bb_rect$y, unitTo = page_units, valueOnly = TRUE)
    new_width <- convertWidth(bb_rect$width,
        unitTo = page_units, valueOnly = TRUE
    )
    new_height <- convertHeight(bb_rect$height,
        unitTo = page_units, valueOnly = TRUE
    )

    # =========================================================================
    # MAKE GROB
    # =========================================================================
    name <- paste0(
        "bb_rect",
        length(grep(
            pattern = "bb_rect",
            x = grid.ls(
                print = FALSE,
                recursive = FALSE
            )
        )) + 1
    )
    rect <- grid.rect(
        x = unit(new_x, page_units),
        y = unit(page_height - new_y, page_units),
        width = unit(new_width, page_units),
        height = unit(new_height, page_units),
        just = bb_rect$just, gp = bb_rect$gp,
        name = name
    )

    # =========================================================================
    # ADD GROB TO OBJECT
    # =========================================================================

    bb_rect$grobs <- rect

    # =========================================================================
    # RETURN OBJECT
    # =========================================================================

    message("bb_rect[", rect$name, "]")
    invisible(bb_rect)
}
